### ISSUE TITLE
Fault-Tolerant Waku Store protocol

### ISSUE THREAD

**Issue by staheri14** [vacp2p/research#64]
# Problem
The current design of store protocol is not fault-tolerant that is when a peer  goes offline, it will have to risk missing all the messages transmitted in the network during its off time. This also means that other peers who rely on that store peer are subject to this message history gap.
We would like to make store protocol more fault-tolerant in such a way that store peers can safely go offline and be guaranteed that when they get back online they would be able to retrieve history of messages from other store peers.

The current issue outlines a minimal solution with some rough implementation details. Further issues will be followed as the solution grows into a more mature state. 

## Problem structure
We can break this problem into three phases:
1. Failure/history gap detection: Peers may fail capturing the intact message history and become inconsistent with other store nodes due to various reasons for example, a peer may miss some of the messages due to a poor network connection, the network topology, a sudden hardware failure, or just simply by going offline. The peer needs to detect its failure and recover from that incomplete state. 
2. Discovery: Once a peer becomes aware of a gap in its message history, it needs to find other peers of the store protocol which have the missing messages. 
3. Transmission: This phase is to fetch the missed messages from another store node(s).

# Initial solution: time window query
In the initial solution, we consider 
1. that a peer intentionally goes offline (so no crash or sudden failure)
2. A basic discovery method where a peer has a hardcoded list of store peers to query from. This can also be an input to the protocol.
3. Messages are transmitted via a direct connection between the two peers.
  
The high level idea is as follows:
A peer has connections to a set L = {P1,...PN} of other store peers (lets call them backup peers).
A peer P in the store protocol keeps track of its on and off time. Lets say P knows he has been off for the time window W.
Once back online, P  loops through all Pi in L:
  -  asks Pi for the messages within W.
  -  Pi checks if it has been online for the duration of W, if yes returns the history of W otherwise, returns empty. For now we consider the exact match but not subset match. That is if Pi is online for Wi which is a subset of W, still the answer to the query is no.

## The implementation path
Base on what I can see in the store protocol, we can build the time window query, more specifically the data transmission part, relying on the pagination feature which is already implemented. Just we need to take into account that:
- the history queries may have no content topics
- the `cursor` as part of pagination info may have no digest

## Potential problems
In this minimal solution, as the discovery is manual, we cannot guarantee 100% data availability. It is because the backup peers may also not have the messages the store peer is looking for. As such, the backup peers must be selected deliberately lets say based on their availability or geographical distribution.

## Immediate improvements to the time window query
- Handling Asynchrony: Peers are asynchronous, so we should account for timing errors. To handle asyncrony, one way is that the querying node can extend its query window  W = [S, E] by some delta value e.g., W = [S-delta , E+delta].
- Enabling subset match: the queried node responds for any time window Wi which is a subset of the queried window W.

# Future Steps
- A more robust and dynamic discovery layer
- More efficient queries: nodes syncing with each other can be done more efficient/reliably with things like Dispersy and bloom filters.
- Verifiability of the query responses: what if queried peers lie about the history of messages.
- Inference attack: querying node have to reveal their  up-down time from which further information can be inferred e.g., religion, location, etc.

**Comment by staheri14** [comment]
@cammellos  Looking forward to your comments and thoughts on this!

**Comment by staheri14** [comment]
> Is there any similar mechanism in waku2? (The assumption is that we use sender time)

In general we do not timestamp messages in waku v2 so there is no filter on the coming messages based on their timestamp (@oskarth is that right?), We only use timestamp in the store protocol and for the pagination sake. Nevertheless, there the timestamp is generated by the receiver and not the sender.

**Comment by staheri14** [comment]
> Another issue that waku/whisper addresses is injection of messages in the stream. Basically you need to "trust" history nodes as generally you would not accept messages older/newer than +/- 20s of your clock, but with history nodes that's necessary, as you'd be receiving older messages. This effectively makes history nodes kind of special nodes and not just any nodes.

I agree with your point! we need to be aware of malicious nodes and take measures. This is also included in the "future step" part of the current issue and will certainly address it as the implementation grows to a more mature state. Thanks!

**Comment by oskarth** [comment]
> l we do not timestamp messages in waku v2 so there is no filter on the coming messages based on their timestamp (@oskarth is that right?)

Correct. This can easily be added to https://rfc.vac.dev/spec/14/ though.

We could introduce it but make it a soft requirement initially w.r.t. store nodes and being pseudo clock synchronized. (This also plays into feedback from store nodes in terms of if a message is accepted or not (being too out of sync), which is a feature I assume the Core app may need.)

**Comment by oskarth** [comment]
@staheri14 what is missing in terms of research/specs for this?
